services/outputhost/kafkaStream.go

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package outputhost
+
+import (
+	"errors"
+	"strconv"
+
+	s "github.com/Shopify/sarama"
+	sc "github.com/bsm/sarama-cluster"
+	"github.com/uber/cherami-server/common"
+	"github.com/uber/cherami-server/stream"
+	"github.com/uber/cherami-thrift/.generated/go/cherami"
+	"github.com/uber/cherami-thrift/.generated/go/store"
+)
+
+// kafkaStream implements the same interface and similar-enough semantics to the Cherami store interface (BStoreOpenReadStreamOutCall)
+
+type kafkaStream struct {
+	creditSemaphore common.UnboundedSemaphore
+	kafkaMsgsCh     <-chan *s.ConsumerMessage
+}
+
+var kafkaErrNilControlFlow = errors.New(`nil or non-positive controlFlow passed to Write()`)
+var kafkaErrClosed = errors.New(`closed`)
+
+/*
+ * BStoreOpenReadStreamOutCall Interface
+ */
+
+// Write grants credits to the Read call
+func (k *kafkaStream) Write(arg *cherami.ControlFlow) error {
+	if arg == nil || arg.GetCredits() <= 0 {
+		return kafkaErrNilControlFlow
+	}
+	k.creditSemaphore.Release(int(arg.GetCredits()))
+	return nil
+}
+
+// Flush is a no-op
+func (k *kafkaStream) Flush() error { return nil }
+
+// Done is a no-op; consumer connection is handled at a higher level
+func (k *kafkaStream) Done() error { return nil }
```

Comment:
I switched the order around, which seems wrong, but it agrees with the future development of  size-based credits.